File: algo-dsa-beginner/02_static-array/python/remove-dup-sorted.py
class Solution:

    def removeDuplicates(self, nums: list[int]) -> int:
        if (len(nums) < 2):
            return len(nums)

        r_head = 1
        w_head = 0

        while (r_head < len(nums)):

            #print(f"r_head:{r_head}, w_head:{w_head}")
            if (nums[r_head] != nums[w_head]):
                nums[w_head + 1] = nums[r_head]
                w_head += 1

            r_head += 1

        del nums[w_head + 1:]
        return w_head + 1


    def removeDuplicates2(self, nums: list[int]) -> int:
        if len(nums) == 0:
            print(f"output: []")
            return 0

        r_head = 1
        w_head = 1

        while r_head < len(nums):
            if nums[r_head] == nums[w_head -1]:
                r_head += 1
            else:
                nums[w_head] = nums[r_head]

                r_head += 1
                w_head += 1

        nums = nums[:w_head]
        print(f"output: {nums}")
        return w_head

ex1 = [1,1,2,3,4]
ex1_out = [1,2,3,4]
ex1_ret = 4

ex2 = [2,10,10,30,30,30]
ex2_out = [2,10,30]
ex2_ret = 3

sol = Solution()
print()

ret = sol.removeDuplicates(ex1)
print(f"expected: {ex1_out}")
print(f"actual: {ex1}")
print(f"ex1 ret = {ret}")
print(f"expected: {ex1_ret}")
print()

ret = sol.removeDuplicates(ex2)
print(f"expected: {ex2_out}")
print(f"actual: {ex2}")
print(f"ex2 ret = {ret}")
print(f"expected: {ex2_ret}")
print()

File: algo-dsa-beginner/02_static-array/cpp/remove-dup-sorted.cpp
#include <iostream>
#include <string>
#include <vector>

template <typename T> using vector = std::vector<T>;
using string = std::string;

template <typename T>
void print_vec(const vector<T>& nums) {
    std::cout << "[ ";
    for (size_t i = 0; i < nums.size(); i++) {
        std::cout << (i == 0 ? "" : ", ") << nums[i];
    }
    std::cout << " ]\n";
}

class Solution {
public:

    int removeDuplicates(vector<int>& nums) {
        // 0 or 1 dont do anything
        if (nums.size() < 2) {
            return nums.size();
        }

        size_t r_head = 1;
        size_t w_head = 1;

        while (r_head < nums.size()) {
            if (nums[w_head - 1] != nums[r_head]) {
                nums[w_head] = nums[r_head];
                ++w_head;
            }
            ++r_head;
        }

        nums.resize(w_head);

        print_vec(nums);
        return w_head;

    }

    int removeDuplicates_old(vector<int>& nums) {
        if (nums.size() == 0) {
            std::cout << "[ ]\n";
            return 0;
        }

        int k = 1;
        size_t r_head = 1;
        size_t w_head = 1;

        while (r_head < nums.size()) {
            if (nums[r_head] == nums[w_head - 1]) {
                r_head++;
            } else {
                nums[w_head] = nums[r_head];
                r_head++;
                w_head++;
                k++;
            }
        }


        nums.resize(w_head);
        print_vec(nums);

        return k;
    }
};

int main() {

    Solution sol;

    std::cout << "\nex1: ";
    vector<int> ex1 = {1,1,2,3,4};
    print_vec(ex1);

    auto ret = sol.removeDuplicates(ex1);
    std::cout << "ex1 expected: [1,2,3,4]\n";
    std::cout << "ex1 ret expected: 4, acutal: " << ret << "\n";

    std::cout << "\nex1: ";
    vector<int> ex2 = {2,10,10,30,30,30};
    print_vec(ex2);
    ret = sol.removeDuplicates(ex2);

    std::cout << "ex2 expected: [2,10,30]\n";
    std::cout << "ex2 ret expected: 3, acutal: " << ret << "\n";

    return 0;
}
